### ISSUE TITLE
[BUG] salt-cloud sends object of type bytes in an event which causes an error

### ISSUE THREAD

**Issue by dkacar-oradian** [saltstack/salt#57349]
**Description**
I tried to watch Salt's event bus while creating a VM with `salt-cloud`. After seeing some events I got an error:
```
Passed invalid arguments: Object of type 'bytes' is not JSON serializable
```
This shouldn't be happening.

**Setup**

I'm using VmWare provider. This is `salt-cloud` invocation:
```
salt-cloud -m stest.map stest-app-01.xdc
```
The same thing happens if I use a profile with `-p` option.

On the salt master I'm watching events like this:
```
salt-run state.event pretty=True
```

**Steps to Reproduce the behavior**

I can see a few events, but then an error is reported and `salt-run` command exits. Here's the abridged output:

```
salt/cloud/stest-app-01.xdc/creating    {
    "_stamp": "2020-05-18T09:05:08.907986",
    "driver": "vmware",
    "event": "starting create",
    "name": "stest-app-01.xdc",
    "profile": "CentOS-7-stest",
    "provider": "vmware-it-soft-ZG:vmware"
}
salt/cloud/stest-app-01.xdc/requesting  {
...
}
salt/cloud/stest-app-01.xdc/deploying   {
...
}
salt/auth       {
    "_stamp": "2020-05-18T09:09:31.149561",
    "act": "accept",
    "id": "stest-app-01.xdc",
    "pub": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAzOLuWir5KQVyJEelcafw\nurygTJ35qrZ0Q3uZlXojWDTQyPqEAOQZMPPG+RJofy36fJIaTISCNTmaMGUe3XtP\n6y2Pa8gs3txkWq14l7NJbBX6KAzZjzRoPI7ujrMrtlXwtdJYvtffR+fFIzdTGzUT\nk1q1SxVfKtB79hrKH2a04dNM9BhQ92Q5r8pG34P5ZpAEN4SNr/+s68sMcx5Sio70\nQHVDLb1WI8jhztDOQWW2VGEWdfvZ1vAlE5N9XmIE1V2eHTjL6b8zcOKEI6gHThNN\n33U6gTtglmYOdZ8JqEAqd9lDv1jimupRFjPRNsa6rqaHaj/voNR9ElRS1DCFhvPb\nqr0jjNbZ6LizcQlMcjt4/WV5UF3qqiY2FGD2IC5oXerduy5AMr/vNJeTk6Cuhs++\nhGlQRf6tx7f8M8V3q5DkdCoyvteyJ0cmg/+THgmlGEfAi81t0H8mwrjrwrQ93uBT\nSQP0w2iO7iamxY8yP1MMh2SeKR2r41mxQn7SMY6wKzCeuw5jfXaM8GnTtLARaTVE\nM3/skOpP1OJPPy4ydUhlLT32nZWNrh47PScD9jmdYk6dxl/Dw2UMY0huOUMUiE7f\nluJfMaPWhmKhlznwT+9n2XLaJuYvB6FBK1uZKKBRBWxvF6+NTWDpfHfjxB69bB5T\nar+Q3DOexEC9O05xfHxS07ECAwEAAQ==\n-----END PUBLIC KEY-----\n",
    "result": true
}
[INFO    ] Runner completed: 20200518090455066356

Passed invalid arguments: Object of type 'bytes' is not JSON serializable

Usage:

    Watch Salt's event bus and block until the given tag is matched

    .. versionadded:: 2014.7.0
    .. versionchanged:: 2019.2.0

...
```

**Expected behavior**

There should be no error reported and I should be able to see all events. This was the case with some of the previous versions, but I don't recall which ones I used to test these things before.

In theory there could be something else sending this binary object, but I've tried a number of times with the different `salt-cloud` options (with or without map file, creating one or more VMs in parallel) and the error always happens in the same place: after `salt/auth` event belonging to the VM being created.

I'm actually trying to debug an issue with `salt/cloud/*/created` event, but I can't see it.

**Versions Report**
<details><summary>salt --versions-report</summary>

```

# salt-cloud --versions-report
Salt Version:
            Salt: 3000.3
 
Dependency Versions:
 Apache Libcloud: 2.2.1
            cffi: Not Installed
        cherrypy: Not Installed
        dateutil: Not Installed
       docker-py: Not Installed
           gitdb: Not Installed
       gitpython: Not Installed
          Jinja2: 2.11.1
         libgit2: Not Installed
        M2Crypto: 0.35.2
            Mako: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.6.2
    mysql-python: Not Installed
       pycparser: Not Installed
        pycrypto: 2.6.1
    pycryptodome: Not Installed
          pygit2: Not Installed
          Python: 3.6.8 (default, Apr  2 2020, 13:34:55)
    python-gnupg: Not Installed
          PyYAML: 3.12
           PyZMQ: 15.3.0
           smmap: Not Installed
         timelib: Not Installed
         Tornado: 4.5.3
             ZMQ: 4.1.4
 
System Versions:
            dist: centos 7.8.2003 Core
          locale: ISO-8859-2
         machine: x86_64
         release: 4.4.169-1.el7.elrepo.x86_64
          system: Linux
         version: CentOS Linux 7.8.2003 Core
```

</details>

**Additional context**

**Comment by garethgreenaway** [comment]
@dkacar-oradian Thanks for the report.  Are you able to check the Salt master logs for anything that happens around the time when this issue happens on the event bus?

**Comment by dkacar-oradian** [comment]
I have master log at the debug level.

This is suspicious (I added line breaks for readability, but it's just one line in the log):

```
2020-05-22 12:58:49,570 [salt.utils.event :736 ][DEBUG   ][7329] Sending event: tag = salt/run/20200522105849533776/new;
data = {'fun': 'runner.vault.generate_token',
'jid': '20200522105849533776',
'user': 'UNKNOWN',
'fun_args': [{'minion_id': 'stest-app-01.xdc',
'signature': b'vvg6KVjBHbcM8YosHBEL0uFrGVKQmeIvn6zhh/1XaZW99Zl8a8g0/wgXFgDtAGBwSCjllgtd6H4ZpJFePJqwtueQvXWM3059jTVoe4IKwVggoSNkDFiyiaiORJH1MFqYf64JVmZGUuYhhDjlcBS7M6x6kRUhp4d3w5kpHdJooH8n8U/hfik2/tu5L3+1lOp0E62Coi1ohOGsZ3DU08sQsTTTBe8SHkV/t/zRh1rHITyLvDHq0EfQTDROloHo7pY3iWWyPPoJk6yPwwHJ/JrnIVN+1n9tGypdlSIU0h16SkIA/EGcdn+L+MnG3/BQ2whycri7KEZzeyvMEmwvz38Oaw==',
 'impersonated_by_master': True}],
'_stamp': '2020-05-22T10:58:49.570672'}
```

I suppose the data in `signature` is binary. And this is the first event in master log after the `salt/auth` event which is the last shown before `salt-run state.event` exits.

And the one after that is:

```
2020-05-22 12:58:49,576 [salt.utils.event :736 ][DEBUG   ][7339] Sending event: tag = salt/run/20200522105323806569/ret;
data = {'fun': 'runner.state.event',
'jid': '20200522105323806569',
'user': 'root',
'fun_args': [{'pretty': True}],
'_stamp': '2020-05-22T10:58:49.575954',
'return': "\nPassed invalid arguments: Object of type 'bytes' is not JSON serializable\n\nUsage:\n\n    Watch Salt's event bus and block until the given tag is matched\n\n ...
```

Master's debug level was propagated to the `salt-run`, so terminal output looks like this:

```
salt/auth       {
    "_stamp": "2020-05-22T10:58:48.395673",
    "act": "accept",
    "id": "stest-app-01.xdc",
    "pub": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAxRELKUcsQAI/Zf4lukyY\nSeshnS61fX7VzhG21aUyDtldRIPYLqu5hnMpdDVxBrSZIBAKZgPFDKl+DpReq/qu\ndt9Bf9Z/Dyc6P05Q/3ggMhUkImsL3vaQVbdOlV8/lQcfQA/wj4vSt1e1VJe4VYhI\n5E1I5dWDUBpLYQrk+T/qB6ioEAwWQmoCO876IQXNo1Hyap76kHH0e+JlO/MvRzhv\nXXwkqqfctMTysmxzqLpX6C/T5QCr56cewHXUVk7b1VCdVGicbgmxhJchEbm7mfdg\nmUeTPv/41DjvAes9gf3USU6VO7jwuhuwGHCyJWwQyiyEupS4HsIoUOttLhofpVHm\nm864Gh8vB/FYmFWJhfY9p+s68Q0UnYPJu9WQQVOowdv3PK6yK0QTulUPFvYLVjWC\nYlWOqGeXEF0ItV1nnI/sb9zVx3wc4Mi0MxcuDKC5gSvOsSecC88YwZqVShePF9Dc\njcpeKs/IBG8rAJ8PVz04bfXJA5OpnODQsPQKk+CigUamRyY+g08UXF/DN+CZd4Vd\nSjgWYhB+4qy9eQtnL57aExq6kk2u/paQpJlgyOe1wrBvDCKNZRR8NtVvxjcNnIwg\nkRIs3NwCXNkVTuAHryMgHZr0rN3at22ie/dNL2KmCn4RKdvuDW0dikP3pX718+3K\nB2/LXD/eC993wFHlfshVBzMCAwEAAQ==\n-----END PUBLIC KEY-----\n",
    "result": true
}
[DEBUG   ] Closing IPCMessageSubscriber instance
[DEBUG   ] LazyLoaded local_cache.prep_jid
[DEBUG   ] Sending event: tag = salt/run/20200522105323806569/ret; data = {'fun': 'runner.state.event', 'jid': '20200522105323806569', 'user': 'root', 'fun_args': [{'pretty': True}], '_stamp': '2020-05-22T10:58:49.575954', 'return': "\nPassed invalid arguments: Object of type 'bytes' is not JSON serializable\n\n ...
[INFO    ] Runner completed: 20200522105323806569
[DEBUG   ] Closing IPCMessageClient instance
[DEBUG   ] Runner return:
...
```

Minion config on the server `stest-app-01.xdc`:
```
grains:
  salt-cloud:
    driver: vmware
    profile: CentOS-7-stest
    provider: vmware-it-soft-ZG:vmware
hash_type: sha256
id: stest-app-01.xdc
log_level: info
master: salt.oradian.ord
pillar_raise_on_missing: true
pillarenv: base
saltenv: base
sls_list:
- serverinit
startup_states: sls
use_superseded:
- module.run
```

Startup state `serverinit` does not access Vault, but some of the pillar files do.

**Comment by garethgreenaway** [comment]
@dkacar-oradian Thanks.  This is potentially helpful as it might be or be related to the vault generate_token command being run.

**Comment by dkacar-oradian** [comment]
For a test server I have disabled pillars which access Vault and created it with:
```
salt-cloud -p stest-dbsnap-01.xdc stest-dbsnap-01.xdc
```
In another terminal I was running:
```
salt-run state.event pretty=True
```
And everything went fine. `salt-run` reported all events and didn't exit on error.

This is the interesting output from `salt-run` command:
```
salt/auth       {
    "_stamp": "2020-05-25T07:05:40.213219",
    "act": "accept",
    "id": "stest-dbsnap-01.xdc",
    "pub": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA4z33LhkoOfSovk4vaQWg\nN70RxS1D+LucdXlqN/Dh2yNr9V6hHzCnnVvuEqY/1nWtRJGG3ALWsjaQVmg8kT82\nJkMchnGjSW/k+3OrFtOXVlKIZuM08Yrn1txteh7TfXTqtCXaOOhbx6vk21ZfyE14\nWi0+YhSrRpVZjKhP3KONPGehctRy7ovIjYR6QabaIC3ncOioB97GnSa4v7n2ppTc\n2p3AXE/MNDh6DD1MktbDQTg8/17vZi9fnrmk5C7f+fZgQoA5wMbuW9jewyPx40+2\nMavTzPUMTQp3ZQHp2/WqmhSKbNKsGhfyMz7v45LO/mUY+zKdXS71rX/CTLlEULXA\nKxm30lLUplL6JmcUNWLxk5yvvZk7Smb/KokA45A4uCyduUYUxw2QyYAG7STCzjKl\nwIQ1WItE19c65K02SP7dhppkAtn9jiZKHrij4BX2k5Jt3sBav+qudLYbQF2P2eeL\ntVveLOHVBudUqqlVBJHE7Zlk3yVIBVNzlJJpMAvEZhrHnDsqkXjI4tuSnv7t9p3G\nEL02durRyS1kutOhZbqrQvPpJSljZaC5jtATHtypFOKI9KTX86+inA0dNPHHakIp\nN6LIOVW7dB5dobVZAAeyRW1wjYkGPCMGbLbSLFCNBIf6/4Ljm9ii0d7Iq+F5WL5U\nGhUxpaqhppSnZm9Yk0Gm/b0CAwEAAQ==\n-----END PUBLIC KEY-----\n",
    "result": true
}
minion/refresh/stest-dbsnap-01.xdc      {
    "Minion data cache refresh": "stest-dbsnap-01.xdc",
    "_stamp": "2020-05-25T07:05:40.662750"
}
minion_start    {
    "_stamp": "2020-05-25T07:05:41.194568",
    "cmd": "_minion_event",
    "data": "Minion stest-dbsnap-01.xdc started at Mon May 25 09:05:40 2020",
    "id": "stest-dbsnap-01.xdc",
    "pretag": null,
    "tag": "minion_start"
}
salt/minion/stest-dbsnap-01.xdc/start   {
    "_stamp": "2020-05-25T07:05:41.210816",
    "cmd": "_minion_event",
    "data": "Minion stest-dbsnap-01.xdc started at Mon May 25 09:05:40 2020",
    "id": "stest-dbsnap-01.xdc",
    "pretag": null,
    "tag": "salt/minion/stest-dbsnap-01.xdc/start"
}
minion/refresh/stest-dbsnap-01.xdc      {
    "Minion data cache refresh": "stest-dbsnap-01.xdc",
    "_stamp": "2020-05-25T07:05:44.717867"
}
salt/job/20200525070546121351/ret/stest-dbsnap-01.xdc   {
    "_stamp": "2020-05-25T07:05:46.133137",
    "arg": [
        [
            "serverinit"
        ]
    ],
    "cmd": "_return",
    "fun": "state.sls",
    "fun_args": [
        [
            "serverinit"
        ]
    ],
   ...
}
salt/auth       {
    "_stamp": "2020-05-25T07:05:50.638748",
    "act": "accept",
    "id": "stest-dbsnap-01.xdc",
    "pub": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA4z33LhkoOfSovk4vaQWg\nN70RxS1D+LucdXlqN/Dh2yNr9V6hHzCnnVvuEqY/1nWtRJGG3ALWsjaQVmg8kT82\nJkMchnGjSW/k+3OrFtOXVlKIZuM08Yrn1txteh7TfXTqtCXaOOhbx6vk21ZfyE14\nWi0+YhSrRpVZjKhP3KONPGehctRy7ovIjYR6QabaIC3ncOioB97GnSa4v7n2ppTc\n2p3AXE/MNDh6DD1MktbDQTg8/17vZi9fnrmk5C7f+fZgQoA5wMbuW9jewyPx40+2\nMavTzPUMTQp3ZQHp2/WqmhSKbNKsGhfyMz7v45LO/mUY+zKdXS71rX/CTLlEULXA\nKxm30lLUplL6JmcUNWLxk5yvvZk7Smb/KokA45A4uCyduUYUxw2QyYAG7STCzjKl\nwIQ1WItE19c65K02SP7dhppkAtn9jiZKHrij4BX2k5Jt3sBav+qudLYbQF2P2eeL\ntVveLOHVBudUqqlVBJHE7Zlk3yVIBVNzlJJpMAvEZhrHnDsqkXjI4tuSnv7t9p3G\nEL02durRyS1kutOhZbqrQvPpJSljZaC5jtATHtypFOKI9KTX86+inA0dNPHHakIp\nN6LIOVW7dB5dobVZAAeyRW1wjYkGPCMGbLbSLFCNBIf6/4Ljm9ii0d7Iq+F5WL5U\nGhUxpaqhppSnZm9Yk0Gm/b0CAwEAAQ==\n-----END PUBLIC KEY-----\n",
    "result": true
}
salt/cloud/stest-dbsnap-01.xdc/deploy_script    {
    "_stamp": "2020-05-25T07:05:50.953863",
    "event": "stest-dbsnap-01.xdc has been deployed at 10.0.54.15",
    "host": "10.0.54.15",
    "name": "stest-dbsnap-01.xdc"
}
minion/refresh/stest-dbsnap-01.xdc      {
    "Minion data cache refresh": "stest-dbsnap-01.xdc",
    "_stamp": "2020-05-25T07:05:51.158831"
}
salt/job/20200525070552241121/ret/stest-dbsnap-01.xdc   {
    "_stamp": "2020-05-25T07:05:52.244073",
    "arg": [
        "salt-minion"
    ],
    "cmd": "_return",
    "fun": "service.restart",
    "fun_args": [
        "salt-minion"
    ],
    "id": "stest-dbsnap-01.xdc",
    "jid": "20200525070552241121",
    "retcode": 0,
    "return": true,
    "tgt": "stest-dbsnap-01.xdc",
    "tgt_type": "glob"
}
salt/cloud/stest-dbsnap-01.xdc/created  {
    "_stamp": "2020-05-25T07:05:52.441441",
    "driver": "vmware",
    "event": "created instance",
    "minion": {
        "grains": {
            "diskconfig": {
                "os": [
                    "/dev/sda"
                ],
                "pgextra": [
                    "/dev/sdb"
                ],
                "pgmirror": [
                    "/dev/sdc",
                    "/dev/sdd",
                    "/dev/sde",
                    "/dev/sdf"
                ]
            }
        },
        "master": "salt.oradian.ord",
        "pillar_raise_on_missing": true,
        "pillarenv": "dave",
        "saltenv": "dave",
        "sls_list": [
            "serverinit"
        ],
        "startup_states": "sls",
        "use_superseded": [
            "module.run"
        ]
    },
    "name": "stest-dbsnap-01.xdc",
    "profile": "stest-dbsnap-01.xdc",
    "provider": "vmware-it-soft-ZG:vmware"
}
salt/run/20200525070553806872/new       {
    "_stamp": "2020-05-25T07:05:53.810748",
    "fun": "runner.salt.cmd",
    "fun_args": [
        {
            "fun": "slack.call_hook",
            "message": "Created `stest-dbsnap-01.xdc` in salt environment `dave`."
        }
    ],
    "jid": "20200525070553806872",
    "user": "Reactor"
}
salt/run/20200525070553806872/ret       {
    "_stamp": "2020-05-25T07:05:54.797627",
    "fun": "runner.salt.cmd",
    "fun_args": [
        {
            "fun": "slack.call_hook",
            "message": "Created `stest-dbsnap-01.xdc` in salt environment `dave`."
        }
    ],
    "jid": "20200525070553806872",
    "return": true,
    "success": true,
    "user": "Reactor"
}
```

**Comment by garethgreenaway** [comment]
@dkacar-oradian Sounds like maybe there is something in the event that is being generated for vault related activity that Salt is able to handle properly but the event runner is not.  This gives us some additional information to try to replicate this and get a fix in. Thanks!

**Comment by andsk8** [comment]
Any update on this?
I have the

**Comment by garethgreenaway** [comment]
Apologies for the delay on this one, while working on another issue and finding the cause, I believe the fix will also resolve this one, the root of the issue is that the data element contains bytes which is incompatible with `salt.utils.json.dumps`, the fix I've been looking at is this:

diff --git a/salt/modules/state.py b/salt/modules/state.py
index 09e03c9981..61ba8c6ea2 100644
--- a/salt/modules/state.py
+++ b/salt/modules/state.py
@@ -2477,7 +2477,7 @@ def event(
                         "{}\t{}".format(  # future lint: blacklisted-function
                             salt.utils.stringutils.to_str(ret["tag"]),
                             salt.utils.json.dumps(
-                                ret["data"],
+                                salt.utils.data.decode(ret["data"]),
                                 sort_keys=pretty,
                                 indent=None if not pretty else 4,
                             ),

Using the salt.utils.data.decode we can ensure that data is in the right format regardless of if that element is a string, a list, or a dictionary. If you're able to, can you test this fix out and see if it resolves the issue for you?

**Comment by gcavelier** [comment]
@garethgreenaway , I had the same issue and using salt.utils.data.decode fixed it.

**Comment by whytewolf** [comment]
this looks like it got missed when #59260 was merged. closing now.
